pkg/tools/clientmap/refcount.go

```diff
@@ -17,73 +17,114 @@
+type RefcountMap struct {
+	m sync.Map
 }
 
-func (r *refcountClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
-	return r.client.Close(ctx, conn)
+// Store sets the value for a key
+// count = 1.
+func (m *RefcountMap) Store(key string, value networkservice.NetworkServiceClient) {
+	m.m.Store(key, &entry{
+		count: 1,
+		value: value,
+	})
 }
 
-// RefcountMap - clientmap.Map wrapped with refcounting.  Each Store,Load,LoadOrStore increments the count,
-//               each Delete decrements the count.  When the count is zero, Delete deletes.
-type RefcountMap struct {
-	Map
-}
+// LoadOrStore returns the existing value for the key if present. Otherwise, it stores and returns the given value. The
+// loaded result is true if the value was loaded, false if stored.
+// store  -->  count = 1
+// load   -->  count += 1
+func (m *RefcountMap) LoadOrStore(key string, value networkservice.NetworkServiceClient) (_ networkservice.NetworkServiceClient, loaded bool) {
+	var raw interface{}
+	raw, loaded = m.m.LoadOrStore(key, &entry{
+		count: 1,
+		value: value,
+	})
+	entry := raw.(*entry)
+	if !loaded {
+		return entry.value, false
+	}
 
-// Store sets the value for a key.
-// Increments the refcount
-func (r *RefcountMap) Store(key string, value networkservice.NetworkServiceClient) {
-	client := &refcountClient{client: value}
-	r.Map.Store(key, client)
-	atomic.AddInt32(&client.count, 1)
-}
+	entry.lock.Lock()
 
-// LoadOrStore returns the existing value for the key if present. Otherwise, it stores and returns the given value. The loaded result is true if the value was loaded, false if stored.
-// Increments the refcount.
-func (r *RefcountMap) LoadOrStore(key string, value networkservice.NetworkServiceClient) (networkservice.NetworkServiceClient, bool) {
-	client := &refcountClient{client: value}
-	rv, loaded := r.Map.LoadOrStore(key, client)
-	if client, ok := rv.(*refcountClient); ok {
-		atomic.AddInt32(&client.count, 1)
-		return client.client, loaded
+	if entry.count == 0 {
+		entry.lock.Unlock()
+		return m.LoadOrStore(key, value)
 	}
-	return rv, loaded
+	entry.count++
+
+	entry.lock.Unlock()
+
+	return entry.value, true
 }
 
-// Load returns the value stored in the map for a key, or nil if no value is present. The ok result indicates whether value was found in the map.
-// Increments refcount.
-func (r *RefcountMap) Load(key string) (networkservice.NetworkServiceClient, bool) {
-	rv, loaded := r.Map.Load(key)
-	if client, ok := rv.(*refcountClient); ok {
-		atomic.AddInt32(&client.count, 1)
-		return client.client, loaded
+// Load returns the value stored in the map for a key, or nil if no value is present. The loaded result indicates
+// whether value was found in the map.
+// count += 1
+func (m *RefcountMap) Load(key string) (_ networkservice.NetworkServiceClient, loaded bool) {
```

Comment:
```suggestion
func (m *RefcountMap) Load(key string) (networkservice.NetworkServiceClient, bool) {
```

Reply:
There are loaded and/or deleted return names for the rest of the methods. For me it looks properly if all methods have the same styling.

Reply:
OK then let's name somehow the client return value to be consistent with styling.

Reply:
Fixed.